pkg/kvstore/memory.go

```diff
@@ -0,0 +1,342 @@
+	clusterName := ""
+	if im.dummy {
+		// If setConfigDummy() was called all clients share the same table.
+		clusterName = "dummy"
+	} else if opts != nil {
+		clusterName = opts.ClusterName
+	}
+	errChan := make(chan error)
+	close(errChan)
+
+	im.mu.Lock()
+	defer im.mu.Unlock()
+	if c, found := im.clients[clusterName]; found {
+		return c, errChan
+	}
+	c := newInMemoryClient(im.db, clusterName)
+	im.clients[clusterName] = c
+	return c, errChan
+}
+
+// setConfig implements backendModule.
+func (i *inMemoryModule) setConfig(logger *slog.Logger, opts map[string]string) error {
+	return nil
+}
+
+// setConfigDummy implements backendModule.
+func (im *inMemoryModule) setConfigDummy() {
+	im.dummy = true
+}
+
+// setExtraConfig implements backendModule.
+func (i *inMemoryModule) setExtraConfig(opts *ExtraOptions) error {
+	return nil
+}
+
+var _ backendModule = &inMemoryModule{}
+
+func newInMemoryClient(db *statedb.DB, clusterName string) *inMemoryClient {
+	table, err := statedb.NewTable(
+		"kvstore-"+clusterName,
+		inMemoryKeyIndex,
+	)
+	if err != nil {
+		panic(err)
+	}
+	if err := db.RegisterTable(table); err != nil {
+		panic(err)
+	}
+	return &inMemoryClient{
+		db:          db,
+		table:       table,
+		clusterName: clusterName,
+	}
+}
+
+type inMemoryObject struct {
+	key   string
+	value []byte
+}
+
+var (
+	inMemoryKeyIndex = statedb.Index[inMemoryObject, string]{
+		Name: "key",
+		FromObject: func(obj inMemoryObject) index.KeySet {
+			return index.NewKeySet(index.String(obj.key))
+		},
+		FromKey:    index.String,
+		FromString: index.FromString,
+		Unique:     true,
+	}
+)
+
+type inMemoryClient struct {
+	db          *statedb.DB
+	table       statedb.RWTable[inMemoryObject]
+	clusterName string
+}
+
+// Close implements BackendOperations.
+func (c *inMemoryClient) Close() {
+}
+
+// Connected implements BackendOperations.
+func (c *inMemoryClient) Connected(ctx context.Context) <-chan error {
+	out := make(chan error)
+	close(out)
+	return out
+}
+
+// CreateOnly implements BackendOperations.
+func (c *inMemoryClient) CreateOnly(ctx context.Context, key string, value []byte, lease bool) (bool, error) {
+	wtxn := c.db.WriteTxn(c.table)
+	defer wtxn.Abort()
+	_, hadOld, _ := c.table.Insert(wtxn, inMemoryObject{
+		key:   key,
+		value: value,
+	})
+	if hadOld {
+		return false, fmt.Errorf("key %q existed", key)
+	}
+	wtxn.Commit()
+	return true, nil
+}
+
+// CreateOnlyIfLocked implements BackendOperations.
+func (c *inMemoryClient) CreateOnlyIfLocked(ctx context.Context, key string, value []byte, lease bool, lock KVLocker) (bool, error) {
+	return c.CreateOnly(ctx, key, value, lease)
+}
+
+// Delete implements BackendOperations.
+func (c *inMemoryClient) Delete(ctx context.Context, key string) error {
+	wtxn := c.db.WriteTxn(c.table)
+	defer wtxn.Abort()
+	_, existed, _ := c.table.Delete(wtxn, inMemoryObject{key: key})
+	if !existed {
+		return fmt.Errorf("key %q did not exist", key)
+	}
```

Comment:
The `Delete` method does not return an error if the key does not exist: https://github.com/joamaki/cilium/blob/pr%2Fjoamaki%2Fcm-service-merger/pkg/kvstore/backend.go#L160-L161